executor/sort.go

```diff
@@ -527,8 +496,174 @@ func (e *TopNExec) doCompaction() error {
+}
+
+func (c *SortedRowContainer) lessRow(rowI, rowJ chunk.Row) bool {
+	for i, colIdx := range c.keyColumns {
+		cmpFunc := c.keyCmpFuncs[i]
+		cmp := cmpFunc(rowI, colIdx, rowJ, colIdx)
+		if c.ByItems[i].Desc {
+			cmp = -cmp
+		}
+		if cmp < 0 {
+			return true
+		} else if cmp > 0 {
+			return false
+		}
+	}
+	return false
+}
+
+// keyColumnsLess is the less function for key columns.
+func (c *SortedRowContainer) keyColumnsLess(i, j int) bool {
+	rowI := c.GetList().GetRow(c.rowPtrs[i])
+	rowJ := c.GetList().GetRow(c.rowPtrs[j])
+	return c.lessRow(rowI, rowJ)
+}
+
+func (c *SortedRowContainer) initPointerAndSort(needLock bool) {
+	if needLock {
+		c.m.RLock()
+		defer c.m.RUnlock()
+	}
+	if c.rowPtrs != nil {
+		return
+	}
+	c.rowPtrs = make([]chunk.RowPtr, 0, c.NumRow())
+	for chkIdx := 0; chkIdx < c.NumChunks(); chkIdx++ {
+		rowChk := c.GetChunk(chkIdx)
+		for rowIdx := 0; rowIdx < rowChk.NumRows(); rowIdx++ {
+			c.rowPtrs = append(c.rowPtrs, chunk.RowPtr{ChkIdx: uint32(chkIdx), RowIdx: uint32(rowIdx)})
+		}
+	}
+	sort.Slice(c.rowPtrs, c.keyColumnsLess)
+	c.GetMemTracker().Consume(int64(8 * cap(c.rowPtrs)))
+}
+
+func (c *SortedRowContainer) sortAndSpillToDisk(needLock bool) (err error) {
+	if needLock {
+		c.m.Lock()
+		defer c.m.Unlock()
+	}
+	c.initPointerAndSort(!needLock)
+	return c.RowContainer.SpillToDisk(needLock)
+}
+
+// Add appends a chunk into the SortedRowContainer.
+func (c *SortedRowContainer) Add(chk *chunk.Chunk) (err error) {
+	c.m.RLock()
+	defer c.m.RUnlock()
+	if c.RowContainer.AlreadySpilled() {
+		return errors.New("The partition is spilled")
+	}
+	return c.RowContainer.Add(chk)
+}
+
+// GetRow returns the row the ptr pointed to.
+func (c *SortedRowContainer) GetRow(ptr chunk.RowPtr) (chunk.Row, error) {
+	c.m.RLock()
+	defer c.m.RUnlock()
+	return c.RowContainer.GetRow(ptr)
+}
+
+// GetRowByIdx returns the row the idx pointed to.
+func (c *SortedRowContainer) GetRowByIdx(idx int) (chunk.Row, error) {
+	c.m.RLock()
+	defer c.m.RUnlock()
+	ptr := c.rowPtrs[idx]
+	return c.GetRow(ptr)
+}
+
+// ActionSpill returns a SortAndSpillDiskAction for sorting and spilling over to disk.
+func (c *SortedRowContainer) ActionSpill() *SortAndSpillDiskAction {
+	c.actionSpill = &SortAndSpillDiskAction{c: c}
+	return c.actionSpill
+}
+
+// SortAndSpillDiskAction implements memory.ActionOnExceed for chunk.List. If
+// the memory quota of a query is exceeded, SortAndSpillDiskAction.Action is
+// triggered.
+type SortAndSpillDiskAction struct {
+	c              *SortedRowContainer
+	fallbackAction memory.ActionOnExceed
+	m              sync.Mutex
+}
+
+// Action sends a signal to trigger sortAndSpillToDisk method of RowContainer
+// and if it is already triggered before, call its fallbackAction.
+func (a *SortAndSpillDiskAction) Action(t *memory.Tracker, trigger *memory.Tracker) {
+	a.m.Lock()
+	defer a.m.Unlock()
+	if a.c.AlreadySpilledSafe() || a.c.GetMemTracker().BytesConsumed() == 0 {
```

Comment:
Maybe create a SortedRowContainer in SortExec , and trigger oom-action in other place.
A spilled SortedRowContainer that consumed 0 bytes is meaningless. It can't add records anymore.